File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.cpp
#include <string>
#include <vector>
#include <string>
#include <cstring>
#include <algorithm>
#include <cctype>
#include <cmath>
using namespace std;

int solution(string str1, string str2) {
    int answer = 0;

    // 두 개씩 자르기
    vector<string> s1; 
    vector<string> s2; 
    
    string str; 
    for(int i = 0; i < str1.size()-1; i++){
        str = "";
        if(('A' <= str1[i] && str1[i] <= 'Z') || ('a' <= str1[i] && str1[i] <= 'z'))
            str += toupper(str1[i]);
        else continue;
        if(('A' <= str1[i+1] && str1[i+1] <= 'Z') || ('a' <= str1[i+1] && str1[i+1] <= 'z'))
            str += toupper(str1[i+1]); 
        else continue;
        s1.push_back(str);
    }
    for(int i = 0; i < str2.size()-1; i++){
        str = "";
        if(('A' <= str2[i] && str2[i] <= 'Z') || ('a' <= str2[i] && str2[i] <= 'z'))
            str += toupper(str2[i]);
        else continue;
        if(('A' <= str2[i+1] && str2[i+1] <= 'Z') || ('a' <= str2[i+1] && str2[i+1] <= 'z'))
            str += toupper(str2[i+1]);
        else continue;
        s2.push_back(str);
    }
    
    // 둘다 정렬
    if(s1.size() == 0 && s2.size() == 0){
        return 65536;
    }else{
        sort(s1.begin(), s1.end());
        sort(s2.begin(), s2.end());
    }
    
    // 둘이 같은거 있으면 교집합에 넣기
    vector<string> inter;
    int s1Idx = 0; int s2Idx = 0; 
    while(s1Idx < s1.size() && s2Idx < s2.size()){
        if(s1[s1Idx] == s2[s2Idx]){
            inter.push_back(s1[s1Idx]); s1Idx++; s2Idx++;
        } else if(s1[s1Idx] < s2[s2Idx]){
            s1Idx++;
        }else{
            s2Idx++;
        }
    }
    
    // 둘이 더하고 교집합꺼 빼기 -> 합집합
    int orSize = s1.size() + s2.size() - inter.size();
    
    // 교집합 크기 / 합집합 크기 * 65536    
    return floor(((double)inter.size() / (double)orSize) * 65536);
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;

class Solution {
    
    private boolean isAlphabet(char c) {
        return ('a' <= c && c <= 'z') ||
               ('A' <= c && c <= 'Z');
    }
    
    public int solution(String str1, String str2) { // 길이 각각 1,000 이하
        int answer = 0;
        
        // 두 문자열을 각각 두 글자씩 끊어서 다중집합의 원소로 만들기 (영문자로 된 쌍만 유효) - 전부 대문자로
        List<String> list1 = new ArrayList<>(); 
        List<String> list2 = new ArrayList<>(); 
        
        for(int i = 0; i < str1.length()-1; i++){
            String s = str1.substring(i, i+2);
            if (isAlphabet(s.charAt(0)) && isAlphabet(s.charAt(1))) {
                list1.add(s.toUpperCase());
            }
        }
        for(int i = 0; i < str2.length()-1; i++){
            String s = str2.substring(i, i+2);
            if (isAlphabet(s.charAt(0)) && isAlphabet(s.charAt(1))) {
                list2.add(s.toUpperCase());
            }
        }
        
        // 집합 사이의 자카드 유사도 구하기
        list1.sort((a, b) -> {return a.compareTo(b);});
        list2.sort((a, b) -> {return a.compareTo(b);});
        List<String> union = getUnion(list1, list2);
        List<String> intersection = getIntersection(list1, list2);
        
        // 자카드 유사도 값은 0에서 1 사이의 실수이므로, 
        // 65536을 곱한 후에 소수점 아래를 버리고 정수부만 출력
        if(intersection.size() == 0 && union.size() == 0) return 65536;
        else {
            return (int)(
                (double)intersection.size() / union.size() * 65536
            );
         }
    }
    
    public List<String> getIntersection(List<String> list1, List<String> list2){
        List<String> intersection = new ArrayList<>(); 
        int oneId = 0; int secId = 0; 
        while(oneId < list1.size() && secId < list2.size()){
            if(list1.get(oneId).equals(list2.get(secId))){
                intersection.add(list1.get(oneId));
                oneId++; secId++;
            }else{
                // 더 작은거 움직이기
                if(list1.get(oneId).compareTo(list2.get(secId)) < 0) oneId++;
                else secId++;
            }
        }
        // 반환
        return intersection;
    }
    
    public List<String> getUnion(List<String> list1, List<String> list2){
        List<String> union = new ArrayList<>(); 
        int oneId = 0; int secId = 0; 
        while(oneId < list1.size() && secId < list2.size()){
            if(list1.get(oneId).equals(list2.get(secId))){
                union.add(list1.get(oneId));
                oneId++; secId++;
            }else{
                // 더 작은거 움직이기
                if(list1.get(oneId).compareTo(list2.get(secId)) < 0) {
                    union.add(list1.get(oneId)); oneId++;
                }
                else {
                    union.add(list2.get(secId)); secId++;
                }
            }
        }
        // 하나만 넘어감
        while(oneId < list1.size()){
            union.add(list1.get(oneId)); oneId++;
        }
        while(secId < list2.size()){
            union.add(list2.get(secId)); secId++;
        }
        // 반환
        return union;
    }
    
}